pam_tacplus.c

```diff
@@ -609,16 +690,290 @@ int pam_sm_close_session (pam_handle_t * pamh, int flags,
+                        else {
+                            _pam_log(LOG_WARNING, "%s: error sending msg=\"%s\", retval=%d",
+                                     __FUNCTION__, conv_msg.msg, retval);
+                        }
+
+                    }
+                    status = PAM_AUTHTOK_ERR;
+                    communicating = 0;
+
+                    _pam_log(LOG_ERR, "chauthtok failed: %d", msg);
+
+                    break;
+
+                case TAC_PLUS_AUTHEN_STATUS_GETDATA:
+                    if (PAM_TAC_DEBUG == (ctrl & PAM_TAC_DEBUG))
+                        syslog(LOG_DEBUG, "tacacs status: TAC_PLUS_AUTHEN_STATUS_GETDATA");
+
+                    if (NULL != conv_msg.msg) {
+                        int retval = -1;
+                        int echo_off = (0x1 == (re.flags & 0x1));
+                        
+                        conv_msg.msg_style = echo_off ? PAM_PROMPT_ECHO_OFF : PAM_PROMPT_ECHO_ON;
+                        retval = converse(pamh, 1, &conv_msg, &resp);
+                        if (PAM_SUCCESS == retval) {
+                            if (PAM_TAC_DEBUG == (ctrl & PAM_TAC_DEBUG)) 
+                                syslog(LOG_DEBUG, "sent msg=\"%s\", resp=\"%s\"",
+                                       conv_msg.msg, resp->resp);
+
+                            if (PAM_TAC_DEBUG == (ctrl & PAM_TAC_DEBUG))
+                                syslog(LOG_DEBUG, "%s: calling tac_cont_send", __FUNCTION__);
+
+                            if (0 > tac_cont_send_seq(tac_fd, resp->resp, re.seq_no + 1)) {
+                                _pam_log(LOG_ERR, "error sending continue req to TACACS+ server");
+                                communicating = 0;
+                            }
+                        }
+                        else {
+                            _pam_log(LOG_WARNING, "%s: error sending msg=\"%s\", retval=%d",
+                                     __FUNCTION__, conv_msg.msg, retval);
+                            communicating = 0;
+                        }
+                    }
+                    else { 
+                        syslog(LOG_ERR, "GETDATA response with no message, returning PAM_TRY_AGAIN");
+                        communicating = 0;
+                    }
+
+                    break;
+
+                case TAC_PLUS_AUTHEN_STATUS_GETUSER:
+                    /* not implemented */
+                    if (ctrl & PAM_TAC_DEBUG)
+                        syslog(LOG_DEBUG, "tacacs status: TAC_PLUS_AUTHEN_STATUS_GETUSER");
+
+                    communicating = 0;
+                    break;
+
+                case TAC_PLUS_AUTHEN_STATUS_GETPASS:
+                    if (ctrl & PAM_TAC_DEBUG)
+                        syslog(LOG_DEBUG, "tacacs status: TAC_PLUS_AUTHEN_STATUS_GETPASS");
+
+                    if (ctrl & PAM_TAC_DEBUG)
+                        syslog(LOG_DEBUG, "%s: calling tac_cont_send", __FUNCTION__);
+
+                    if (tac_cont_send(tac_fd, pass) < 0) {
```

Comment:
What happen if "pass" is null?
"pass" could be null if our lib pam_tacplus is placed in the top of the PAM stack.
As the result, our lib pam_tacplus will send the package without password field.